runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,158 @@
+local M = {}
+local util = require "vim.lsp.util"
+local ns = vim.api.nvim_create_namespace("lsp-semantic-tokens")
+
+local semantic_tokens = {}
+
+local function modifiers_from_number(x, modifiers_table)
+  local function get_bit(n, k)
+    -- Based on/from https://stackoverflow.com/a/26230537
+    -- (n >> k) & 1
+    return bit.band(bit.rshift(n, k), 1)
+  end
+
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+M.token_map = {
+  type = "LspType",
+  class = "LspClass",
+  enum = "LspEnum",
+  interface = "LspInterface",
+  struct = "LspStruct",
+  typeParameter = "LspTypeParameter",
+  parameter = "LspParameter",
+  variable = "LspVariable",
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+local function highlight(buf, token, hl)
```

Comment:
There's already a highlight function for tokens in https://github.com/neovim/neovim/blob/2afbce7651f79b0626ebeae3688274ce18ac2920/runtime/lua/vim/lsp/util.lua#L1493-L1507

Can this be refactored to utilize common code?

Reply:
I think this is difficult because the tokens don't use the range schema (pretty unusual for LSP). I might try to convert the tokens to ranges but I don't think the benefit would be big.

Reply:
Is now a one-liner anyways.